net/http/authn/authn.go

```diff
@@ -72,8 +80,14 @@ func (a *API) Authenticate(req *http.Request) (*http.Request, error) {
 
 	// if there is no authentication at all, we return an "unauthenticated" error,
 	// which may be helpful when debugging
-	if len(X509Certs(ctx)) < 1 && err != nil {
-		return req, errors.New("unauthenticated")
+	if len(X509Certs(ctx)) < 1 && Token(ctx) == "" {
+		err := errors.New("unauthenticated")
+		if len(authnErrors) > 0 {
+			err = errors.WithDetailf(err, "Invalid credentials: %s", strings.Join(authnErrors, "; "))
```

Comment:
Seems like it's a bigger issue of how we want to present errors and what conventions should we use. There are quite a few places where the error details are complete sentences.

This is a little different than the situation described in that code-review-comment because the error detail is a specific value in our already structured error representation. There's no expectation that it will be modified with additional context. Maybe we should look at what we want exposed to the end user and work back from there.